File: examples/python/wekeza_sdk/payments.py
"""
Wekeza API Payments Module
Handles payment initiation and tracking
"""

import requests
import secrets
import time
from typing import Dict, Any, Optional


class WekezaPayments:
    """Handles payment-related API calls"""
    
    def __init__(self, config: Dict[str, str], auth):
        self.base_url = config['base_url']
        self.auth = auth
    
    def _get_headers(self, additional_headers: Optional[Dict[str, str]] = None) -> Dict[str, str]:
        """Get authenticated headers"""
        token = self.auth.get_access_token()
        headers = {
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        }
        if additional_headers:
            headers.update(additional_headers)
        return headers
    
    @staticmethod
    def generate_idempotency_key() -> str:
        """Generate a unique idempotency key"""
        return f"payment_{int(time.time())}_{secrets.token_hex(8)}"
    
    def initiate_payment(self, payment_data: Dict[str, Any], idempotency_key: Optional[str] = None) -> Dict[str, Any]:
        """
        Initiate a payment
        
        Args:
            payment_data: Payment details
            idempotency_key: Optional idempotency key
            
        Returns:
            Dict containing payment response
        """
        try:
            if not idempotency_key:
                idempotency_key = self.generate_idempotency_key()
            
            response = requests.post(
                f"{self.base_url}/payments",
                json=payment_data,
                headers=self._get_headers({'Idempotency-Key': idempotency_key})
            )
            response.raise_for_status()
            return response.json()
        except Exception as e:
            raise self._handle_error(e)
    
    def get_payment(self, payment_id: str) -> Dict[str, Any]:
        """
        Get payment details
        
        Args:
            payment_id: Payment ID
            
        Returns:
            Dict containing payment details
        """
        try:
            response = requests.get(
                f"{self.base_url}/payments/{payment_id}",
                headers=self._get_headers()
            )
            response.raise_for_status()
            return response.json()
        except Exception as e:
            raise self._handle_error(e)
    
    def get_payment_status(self, payment_id: str) -> Dict[str, Any]:
        """
        Get payment status
        
        Args:
            payment_id: Payment ID
            
        Returns:
            Dict containing payment status
        """
        try:
            response = requests.get(
                f"{self.base_url}/payments/{payment_id}/status",
                headers=self._get_headers()
            )
            response.raise_for_status()
            return response.json()
        except Exception as e:
            raise self._handle_error(e)
    
    def list_payments(self, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """
        List payments
        
        Args:
            params: Query parameters (status, fromDate, toDate, page, limit)
            
        Returns:
            Dict containing payment list
        """
        try:
            response = requests.get(
                f"{self.base_url}/payments",
                headers=self._get_headers(),
                params=params or {}
            )
            response.raise_for_status()
            return response.json()
        except Exception as e:
            raise self._handle_error(e)
    
    def cancel_payment(self, payment_id: str, reason: str) -> Dict[str, Any]:
        """
        Cancel a payment
        
        Args:
            payment_id: Payment ID
            reason: Cancellation reason
            
        Returns:
            Dict containing cancellation response
        """
        try:
            response = requests.post(
                f"{self.base_url}/payments/{payment_id}/cancel",
                json={'reason': reason},
                headers=self._get_headers()
            )
            response.raise_for_status()
            return response.json()
        except Exception as e:
            raise self._handle_error(e)
    
    def mpesa_stk_push(self, mpesa_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Initiate M-Pesa STK Push
        
        Args:
            mpesa_data: M-Pesa payment details
            
        Returns:
            Dict containing M-Pesa response
        """
        try:
            response = requests.post(
                f"{self.base_url}/payments/mpesa/stk-push",
                json=mpesa_data,
                headers=self._get_headers()
            )
            response.raise_for_status()
            return response.json()
        except Exception as e:
            raise self._handle_error(e)
    
    def _handle_error(self, error: Exception) -> Exception:
        """Handle API errors"""
        if isinstance(error, requests.HTTPError):
            response = error.response
            try:
                data = response.json()
                message = data.get('message') or data.get('error') or 'Unknown error'
            except:
                message = response.text or 'Unknown error'
            return Exception(f"API Error ({response.status_code}): {message}")
        elif isinstance(error, requests.RequestException):
            return Exception(f"Network error: {str(error)}")
        else:
            return Exception(f"Request error: {str(error)}")

File: examples/javascript/src/payments.js
/**
 * Wekeza API Payments Module
 * Handles payment initiation and tracking
 */

const axios = require('axios');
const crypto = require('crypto');

class WekezaPayments {
  constructor(config, auth) {
    this.baseUrl = config.baseUrl;
    this.auth = auth;
  }

  /**
   * Get authenticated axios instance
   * @returns {Promise<object>} Configured axios instance
   */
  async getClient() {
    const token = await this.auth.getAccessToken();
    return axios.create({
      baseURL: this.baseUrl,
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json'
      }
    });
  }

  /**
   * Generate idempotency key
   * @returns {string} Unique idempotency key
   */
  generateIdempotencyKey() {
    return `payment_${Date.now()}_${crypto.randomBytes(8).toString('hex')}`;
  }

  /**
   * Initiate a payment
   * @param {object} paymentData - Payment details
   * @param {object} options - Additional options (idempotencyKey)
   * @returns {Promise<object>} Payment response
   */
  async initiatePayment(paymentData, options = {}) {
    try {
      const client = await this.getClient();
      const idempotencyKey = options.idempotencyKey || this.generateIdempotencyKey();
      
      const response = await client.post('/payments', paymentData, {
        headers: {
          'Idempotency-Key': idempotencyKey
        }
      });
      
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Get payment details
   * @param {string} paymentId - Payment ID
   * @returns {Promise<object>} Payment details
   */
  async getPayment(paymentId) {
    try {
      const client = await this.getClient();
      const response = await client.get(`/payments/${paymentId}`);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Get payment status
   * @param {string} paymentId - Payment ID
   * @returns {Promise<object>} Payment status
   */
  async getPaymentStatus(paymentId) {
    try {
      const client = await this.getClient();
      const response = await client.get(`/payments/${paymentId}/status`);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * List payments
   * @param {object} params - Query parameters (status, fromDate, toDate, page, limit)
   * @returns {Promise<object>} Payment list
   */
  async listPayments(params = {}) {
    try {
      const client = await this.getClient();
      const response = await client.get('/payments', { params });
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Cancel a payment
   * @param {string} paymentId - Payment ID
   * @param {string} reason - Cancellation reason
   * @returns {Promise<object>} Cancellation response
   */
  async cancelPayment(paymentId, reason) {
    try {
      const client = await this.getClient();
      const response = await client.post(`/payments/${paymentId}/cancel`, { reason });
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Initiate M-Pesa STK Push
   * @param {object} mpesaData - M-Pesa payment details
   * @returns {Promise<object>} M-Pesa response
   */
  async mpesaStkPush(mpesaData) {
    try {
      const client = await this.getClient();
      const response = await client.post('/payments/mpesa/stk-push', mpesaData);
      return response.data;
    } catch (error) {
      throw this.handleError(error);
    }
  }

  /**
   * Handle API errors
   * @param {Error} error - Axios error
   * @returns {Error} Formatted error
   */
  handleError(error) {
    if (error.response) {
      const { status, data } = error.response;
      return new Error(
        `API Error (${status}): ${data.message || data.error || 'Unknown error'}`
      );
    } else if (error.request) {
      return new Error('Network error: No response received from server');
    } else {
      return new Error(`Request error: ${error.message}`);
    }
  }
}

module.exports = WekezaPayments;
